concordium-consensus/tests/consensus/ConcordiumTests/LeaderElectionTest.hs

```diff
@@ -0,0 +1,147 @@
+    parentInNextEpoch :: Slot <- fromIntegral <$> choose (nextEpoch + 1, nextEpoch + fromIntegral epochLength-1) -- For test cases where the parent of the pending block in the same epoch as the target slot
+    let vrfKP = fst $ VRF.randomKeyPair (mkStdGen 0)
+        bn = VRF.prove vrfKP "3"
+        seedStateSlot' = fromIntegral nextEpoch + 1
+        ss' = updateSeedState seedStateSlot' bn ss -- update the seedstate for a slot in the next epoch
+        seedStateSlot'' = seedStateSlot + 1
+        ss'' = updateSeedState seedStateSlot'' bn ss -- update the seedstate for a slot in the same epoch
+    slotToPredictFor :: Slot <- fromIntegral <$> choose (fromIntegral parentInNextEpoch, nextEpoch + fromIntegral epochLength-1) -- the target slot in the next epoch
+
+    return $ conjoin [
+            counterexample "Without knowledge of pending block's parent" $ predictLeadershipElectionNonce ss seedStateSlot Nothing slotToPredictFor === Just [computeLeadershipElectionNonce ss slotToPredictFor, computeLeadershipElectionNonce ss' slotToPredictFor],
+            counterexample "With parent in same epoch as the last finalized block" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInThisEpoch) slotToPredictFor === Just [computeLeadershipElectionNonce ss slotToPredictFor],
+            counterexample "With parent in the same epoch as the target slot" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInNextEpoch) slotToPredictFor === Just [computeLeadershipElectionNonce ss' slotToPredictFor],
+            singletonElementOfProperty (predictLeadershipElectionNonce ss seedStateSlot Nothing slotToPredictFor) (predictLeadershipElectionNonce ss' seedStateSlot' Nothing slotToPredictFor),
+            counterexample "Prediction invariant under seedstate update, with parent in same epoch as the last finalized block" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInThisEpoch) slotToPredictFor === predictLeadershipElectionNonce ss'' seedStateSlot'' (Just parentInThisEpoch) slotToPredictFor,
+            counterexample "Prediction invariant under seedstate update across an epoch, with parent in the same epoch as the target slot" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInNextEpoch) slotToPredictFor === predictLeadershipElectionNonce ss' seedStateSlot' (Just parentInNextEpoch) slotToPredictFor,
+            counterexample "Prediction invariant under seedstate update in same epoch, with parent in the same epoch as the target slot" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInNextEpoch) slotToPredictFor === predictLeadershipElectionNonce ss'' seedStateSlot'' (Just parentInNextEpoch) slotToPredictFor
+        ]
+
+-- Tests cases where the leadership election nonce cannot be predicted, either beceause the seedstate slot is too early (i.e. in the first 2/3 of the epoch), or
```

Comment:
```suggestion
-- Tests cases where the leadership election nonce cannot be predicted, either because the seedstate slot is too early (i.e. in the first 2/3 of the epoch), or
```